### ISSUE TITLE
Misleading edge.From / edge.To semantics

### ISSUE THREAD

**Issue by phanalpha** [ent/ent#3849]
The document states a schema that defines an edge using the `edge.To` builder owns the relation. This is obviously, as a `edge.From` requires an extra `Ref`.

By 'owning', a foreign key is expected to be included in the corresponding table. But as the example (https://entgo.io/docs/schema-edges/#o2o-two-types) shows, the edge field is actually included in the `From` side.

For O2M situation, we can switch `From` and `To` on either sides, as a `Unique` clarify which is which. We can also declare only `To` on the M side, to avoid unnecessary (`From`) field on the O side. And it works almost as expected for the O2M situation.

But things get complicated when it comes to the O2O situation. If we have both `From` and `To` declared, the `From` side owns the foreign key. Since `From` is optional, we could remove it. And then, the foreign key goes to the `To` side. (Of course, there's only `To` declared.)

In my opinion, removing an optional From should not alter the schema, and the 'owning' semantics seem somewhat weird.

- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/ent/ent/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

```go
// foreign_key "roles_users_role"
func (Role) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("user", User.Type).
			Unique().
			Ref("role"),
	}
}

func (User) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("role", Role.Type).
			Unique(),
	}
}
```

And after commenting the `Role` side out,

```
func (Role) Edges() []ent.Edge {
	return []ent.Edge{
		//edge.From("user", User.Type).
		//	Unique().
		//	Ref("role"),
	}
}

// foreign_key "users_roles_role"
func (User) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("role", Role.Type).
			Unique(),
	}
}
```

## Expected Behavior 🤔

Removing a From should not alter the behavior (or schema). It would be even better to align the 'owning' semantics. Or at the very least, document the design decision somewhere.

## Steps to Reproduce 🕹

It seems to be a feature for now.

## Your Environment 🌎

| Tech        | Version |
| ----------- | ------- |
| Go          | 1.21.4  |
| Ent         | v0.12.5 |
| Database    | MySQL   |
| Driver      | https://github.com/go-sql-driver/mysql        |

**Comment by a8m** [comment]
> The document states a schema that defines an edge using the `edge.To` builder owns the relation. This is obviously, as a `edge.From` requires an extra `Ref`.

"Edge owner" in ent does not mean the type/edge table holds the foreign key - it means the other side is optional. Hence, removing the `edge.From` should not affect the foreign key. 

SQL is just one implementation of ent, but the schema can be possibly implemented on a different storage-driver, like we have the experiment for Gremlin. 

Closing as the API cannot be changed, but you're welcome to use [this doc](https://entgo.io/docs/schema-edges#relationship) for guidance.

**Comment by leejuyuu** [comment]
> Hence, removing the edge.From should not affect the foreign key.

**Comment by leejuyuu** [comment]
> Hence, removing the edge.From should not affect the foreign key.

@a8m I think this is what this issue is about. I reproduced @phanalpha's example and made a test repo. Please see that in [this commit](https://github.com/leejuyuu/ent_3849/commit/edad927743e24151dec8acfd68ba6b0acf0cba24#diff-b53ffa1cbc91fb41ec5654f0a79fce7bc6bab93e6d3123f55fd04ef97ab3d235), removing the `edge.From` resulted in the foreign key changed to the other table in the `ent/migrate/schema.go`.

**Comment by phanalpha** [comment]
If we want to declare on only one (child) side, `To` is what we can rely on, which violates the document (and/or tradition). Is this a feature or an accidental behavior?

**Comment by amoiseev** [comment]
Removing `edge.From` works like "default" `edge.From` without `.Unique()`, and M cardinality is assumed.

So looks like `edge.From` is not really optional: you lose control over cardinality and get larger cardinality by default.
If `edge.To` is also not `.Unique()` then an intermediary join table is created for M2M.

Probably a warning during code generation should be added for this?

**Comment by leejuyuu** [comment]
I agree. This behavior should be mentioned in a warning or the docs if it is intentional.

**Comment by phanalpha** [comment]
Say if there is an warning, there should also be a flag to turn it off.

There are actually two targets we'd like to achieve. The first one is to model the relationships, and the second is to instruct code generation. For the first one, a declaration on either side should work; but for the second one, the code is generated if and only if the corresponding `from` or `to` presents.

Thus to preserve current design, shall we a) enforce the `from` / `to` always coming in pair and b) set a new flag up to inhibit code generation if intended?

**Comment by amoiseev** [comment]
Interestingly, `.StorageKey()` is explicitly designed for one-sided configuration on the `edge.To` side.

https://entgo.io/docs/schema-edges#storagekey

># StorageKey
> By default, Ent configures edge storage-keys by the edge-owner (the schema that holds the edge.To), and not the by back-reference (edge.From). This is because back-references are optional and can be removed.

A warning seems like a good starting point:
- without `edge.From()` one might get M2M with an intermediary table, hardly expected (only makes sense for Group-User and the like, most hierarchies are single-parent)
- `edge.From().Unique()` is not default/zero configuration, should not be dropped

And a no-codegen flag looks good too, but I personally thought about dropping `edge.From` because I thought it's redundant (the configuration I have is the default, can go zero configuration). But `edge.From().Unique()`is not the default.
